From 5ef709a6c4f62a85115bea822ec3731baf01d707 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:39:30 -0500
Subject: [drm] dp: retry AUX transactions 32 times (v1.1)

Message-id: <1418852380-13061-80-git-send-email-rclark@redhat.com>
Patchwork-id: 102224
O-Subject: [RHEL7 drm 79/89] drm/dp: retry AUX transactions 32 times (v1.1)
Bugzilla: 1173317
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1173317
Upstream: on drm-next

commit 19a93f042fc241ecdf98543cedfe7c171f8cdf53

Author:     Dave Airlie <airlied@redhat.com>
AuthorDate: Wed Nov 26 13:13:09 2014 +1000
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Mon Dec 15 14:43:38 2014 -0500

    drm/dp: retry AUX transactions 32 times (v1.1)

    At least on two MST devices I've tested with, when
    they are link training downstream, they are totally
    unable to handle aux ch msgs, so they defer like nuts.
    I tried 16, it wasn't enough, 32 seems better.

    This fixes one Dell 4k monitor and one of the
    MST hubs.

    v1.1: fixup comment (Tom).

    Acked-by: Alex Deucher <alexander.deucher@amd.com>
    Cc: stable@vger.kernel.org
    Signed-off-by: Dave Airlie <airlied@redhat.com>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/drm_dp_helper.c b/drivers/gpu/drm/drm_dp_helper.c
index 08e33b8..9f59c90 100644
--- a/drivers/gpu/drm/drm_dp_helper.c
+++ b/drivers/gpu/drm/drm_dp_helper.c
@@ -378,10 +378,11 @@ static int drm_dp_dpcd_access(struct drm_dp_aux *aux, u8 request,
 
  /*
   * The specification doesn't give any recommendation on how often to
-  * retry native transactions, so retry 7 times like for I2C-over-AUX
-  * transactions.
+  * retry native transactions. We used to retry 7 times like for
+  * aux i2c transactions but real world devices this wasn't
+  * sufficient, bump to 32 which makes Dell 4k monitors happier.
   */
- for (retry = 0; retry < 7; retry++) {
+ for (retry = 0; retry < 32; retry++) {
 
   mutex_lock(&aux->hw_mutex);
   err = aux->transfer(aux, &msg);
-- 
1.7.1